src/app/access-requests/features/access-request-duration-edit/access-request-duration-edit.component.ts

```diff
@@ -0,0 +1,327 @@
+
+    const newUntilMaxDate = new Date(date);
+    newUntilMaxDate.setHours(23, 59, 59, 999);
+    newUntilMaxDate.setDate(date.getDate() + Math.round(this.maxDays * this.maxExtend));
+    this.maxUntilDate = newUntilMaxDate;
+  }
+
+  /**
+   * Update the range for the from date based on the until date
+   * @param date - The until date to update the range for
+   */
+  updateAccessStartRanges(date: Date): void {
+    const currentDate = new Date(this.todayMidnight);
+
+    const newFromMinDate = new Date();
+    const dateMinusGrantMaxDays = new Date(date);
+    dateMinusGrantMaxDays.setDate(date.getDate() - this.maxDays * this.maxExtend);
+    if (currentDate < dateMinusGrantMaxDays)
+      newFromMinDate.setTime(dateMinusGrantMaxDays.getTime());
+    else newFromMinDate.setTime(currentDate.getTime());
+
+    const newFromMaxDate = new Date();
+    const dateMinusOneDay = new Date(date);
+    dateMinusOneDay.setHours(0, 0, 0, 0);
+    newFromMaxDate.setTime(dateMinusOneDay.getTime());
+
+    this.minFromDate = newFromMinDate;
+    this.maxFromDate = newFromMaxDate;
+  }
+
+  /**
+   * Validate a date against a given interval
+   * @param inDate - The date to validate
+   * @param min - The minimum date
+   * @param max - The maximum date
+   * @returns a validation error with either minDate or maxDate set or null
+   */
+  #validateDateAgainstMinAndMax(
+    inDate: Date,
+    min: Date,
+    max: Date,
+  ): ValidationErrors | null {
+    if (!inDate) return { invalid: true };
+    if (inDate < min) {
+      return { minDate: true };
+    } else if (inDate > max) {
+      return { maxDate: true };
+    }
+    return null;
+  }
+
+  /**
+   * Get an error message for a given validation state
+   * @param name - The name of the date (e.g. start or end)
+   * @param error - The validation error with minDate, maxDate or invalid set
+   * @returns an error message as a string
+   */
+  getErrorMessageForValidationState(
+    name: string,
+    error: ValidationErrors | null,
+  ): string {
+    if (!error) {
+      return '';
+    } else if (error['invalid']) {
+      return 'Please choose a valid ' + name + ' date.';
+    } else if (error['minDate']) {
+      return 'Please choose a later ' + name + ' date.';
+    } else if (error['maxDate']) {
+      return 'Please choose an earlier ' + name + ' date.';
+    } else {
+      return 'Invalid ' + name + ' date.';
+    }
+  }
+
+  /**
+   * Validate a Form Control against the constraints of the from date
+   * @param control The form Control to validate
+   * @returns a validation error or null
+   */
+  #fromDateValidator = (
+    control: AbstractControl<Date, Date>,
+  ): ValidationErrors | null => {
+    const ret = this.#validateDateAgainstMinAndMax(
+      control.value,
+      this.minFromDate,
+      this.maxFromDate,
+    );
+    this.fromFieldErrorMessage.set(
+      this.getErrorMessageForValidationState('start', ret),
+    );
+    return ret;
+  };
```

Comment:
I think you can merge these two (just a draft)
```suggestion
  #DateValidator = (
    control: AbstractControl<Date, Date>,
    from: boolean,
  ): ValidationErrors | null => {
const [min, max, field, name] = from 
  ? [this.minFromDate, this.maxFromDate, this.fromFieldErrorMessage, 'start'] 
  : [this.minUntilDate, this.maxUntilDate, this.untilFieldErrorMessage, 'end'];
    const ret = this.#validateDateAgainstMinAndMax(
      control.value,
      min,
      max,
    );
    field.set(
      this.getErrorMessageForValidationState(name, ret),
    );
    return ret;
  };
```

Reply:
Also: Logic like this should be tested.